EasyScore: Implementation of proposal in #419

0xfe opened the pull request:
This is an implementation of the notation language proposed in #419. There are a few parts to this:

* `parser.js` - A simple context free parser framework.
* `easyscore.js` - Implementation of the EasyScore grammar. This includes:
   * `Grammar` - Uses the above parser framework to define a syntax for the language.
   * `Builder` - Builds VexFlow elements based on the parser output.
   * `EasyScore` - Helper class that connects `Builder`, `Parser` to create notation.
* `factory.js` - Updated to build `EasyScore` instances.
* Tests available in `tests/parser_tests.js` and `tests/easy_score.js`.

Silverwolf90 commented:
This looks great! I'll put the nitpicky stuff inline.

Do you have any ideas around adding modifiers to notes in general? How do they fit (if at all)?

Silverwolf90 commented on src/easyscore.js:
```diff
@@ -0,0 +1,264 @@
+    return { expect: [this.PIECE, this.PIECES, this.EOL] };
+  }
+  PIECE()  {
+    return { expect: [this.CHORDORNOTE, this.PARAMS],
```
Throughout this file I'd put first key on newlines:

```javascript
{
  CHORD()  {
    return {
      expect: [this.LPAREN, this.NOTES, this.RPAREN],
      run: (state) => this.builder.addChord(state.matches[1])
    };
  },
  SINGLENOTE()   {
    return {
      expect: [this.NOTENAME, this.ACCIDENTAL, this.OCTAVE],
      run: (state) => this.builder.addSingleNote(state.matches[0], state.matches[1], state.matches[2])
    };
  }
}
```

0xfe replied:
Done for all the `expect` rules. I left `token` rules vertically aligned because they're more concise and easier to read that way.

Silverwolf90 commented on src/easyscore.js:
```diff
@@ -0,0 +1,264 @@
+    // Build the note.
+    const keys = this.piece.chord.map(note => note.key + '/' + note.octave);
+    const accs = this.piece.chord.map(note => note.acc || null);
+    const note = this.factory.StaveNote(
```
Can we reformat this to:
```javascript
const note = this.factory.StaveNote({
  keys,
  duration: this.piece.duration,
  dots: this.piece.dots,
  auto_stem: autoStem,
  type: this.piece.type,
});
```

0xfe replied:
Done.

Silverwolf90 commented on src/easyscore.js:
```diff
@@ -0,0 +1,264 @@
+    this.piece.options[key] = value;
+  }
+
+  addNote(key, acc, octave) {
```
Could we move from `acc` to `accid`?

0xfe replied:
Done.

Silverwolf90 commented on src/parser.js:
```diff
@@ -0,0 +1,189 @@
+  // Look for `token` in this.line[this.pos], and return success
+  // if one is found. `token` is specified as a regular expression.
+  matchToken(token, noSpace = false) {
+    const re = noSpace ? new RegExp('^((' + token + '))') : new RegExp('^((' + token + ')\\s*)');
```
`re` is a bit cryptic

0xfe replied:
`re` is a pretty standard abbreviation for `regular expression` (across many languages.)

Silverwolf90 replied:
I guess I've never come across it in JS. How about just `regex` then, as lower-casing the constructor for an instance name is pretty conventional.

Silverwolf90 replied:
Or possibly just `regex` which isn't so rigorously consistent, but a bit clearer (at least to me) than `re`.

Silverwolf90 commented:
Also the tests could still use that cleanup.

0xfe commented:
Addressed all your comments.

Re: modifiers, the only things that's part of the language are accidentals and dots. In VexTab I had specific syntax for all kinds of modifiers and it got messy. Here it would be trivial to add a keyval pair for specific types of modifiers, e.g., `c5/q[articulations='staccatto, upstroke']`.

0xfe commented:
Okay, renamed `re` to `regexp`. I don't have a strong opinion about it. :-)

Silverwolf90 commented:
Cool, that's how I was anticipating they would work. I think that's very straightforward. But this does imply that key/value pairs provided are not necessarily mirrored as key/values on the constructed object, they simply act as patterns to match against and then execute some behavior. This seems like a great place to provide a hook.

0xfe commented:
Yes, the only attribute that you can set on the object right now via the
keyvals is 'id'

On Sat, Aug 20, 2016 at 8:46 AM, Cyril Silverman <notifications@github.com>
wrote:

> Cool, that's how I was anticipating they would work. I think that's very
> straightforward. But this does imply that key/value pairs provided are not
> necessarily mirrored as key/values on the constructed object, they simply
> act as patterns to match against and then execute some behavior. This seems
> like a great place to provide a hook.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/0xfe/vexflow/pull/424#issuecomment-241198272>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOuk70Z-M1WYZ4MLWAzWwdbB8h3FF3Iks5qhvcZgaJpZM4Jo9qD>
> .
>



-- 
Mohit Muthanna [mohit (at) muthanna (uhuh) com]

Silverwolf90 commented on src/parser.js:
```diff
@@ -60,9 +60,10 @@ export class Parser {
   // if one is found. `token` is specified as a regular expression.
   matchToken(token, noSpace = false) {
-    const re = noSpace ? new RegExp('^((' + token + '))') : new RegExp('^((' + token + ')\\s*)');
+    const regexp = noSpace ? new RegExp('^((' + token + '))') :
```
Sorry more nitpicking :)

I find that multi-line ternary are clearer when formatted like this:

```javascript
const regexp = noSpace
  ? new RegExp('^((' + token + '))')
  : new RegExp('^((' + token + ')\\s*)');
```

0xfe replied:
Yes, agreed. Reformatted.

On Sat, Aug 20, 2016 at 8:59 AM, Cyril Silverman <notifications@github.com>
wrote:

> In src/parser.js
> <https://github.com/0xfe/vexflow/pull/424#discussion_r75578219>:
>
> > @@ -60,9 +60,10 @@ export class Parser {
> >    // Look for `token` in this.line[this.pos], and return success
> >    // if one is found. `token` is specified as a regular expression.
> >    matchToken(token, noSpace = false) {
> > -    const re = noSpace ? new RegExp('^((' + token + '))') : new RegExp('^((' + token + ')\\s*)');
> > +    const regexp = noSpace ? new RegExp('^((' + token + '))') :
>
> Sorry more nitpicking :)
>
> I find that multi-line ternary are clearer when formatted like this:
>
> const regexp = noSpace
>   ? new RegExp('^((' + token + '))')
>   : new RegExp('^((' + token + ')\\s*)');
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/0xfe/vexflow/pull/424/files/e574a4fe89334b6b7b350ba68d7de417eb06607a..6c60b81b2df2db46b616194080d28f2bb1968ad7#r75578219>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOuk4vhafxEcvJwYAMlFdWy06zdH735ks5qhvoygaJpZM4Jo9qD>
> .
>



-- 
Mohit Muthanna [mohit (at) muthanna (uhuh) com]

Silverwolf90 commented:
I noticed in the image you included in https://github.com/0xfe/vexflow/issues/419#issuecomment-241169003 the stave lines are rendering over the noteheads.

0xfe commented:
Good catch. Fixed.

<img width="554" alt="screen shot 2016-08-20 at 1 08 37 pm" src="https://cloud.githubusercontent.com/assets/241299/17832712/416c6fee-66d7-11e6-9e26-9c84476b7fc0.png">
<img width="581" alt="screen shot 2016-08-20 at 1 08 32 pm" src="https://cloud.githubusercontent.com/assets/241299/17832713/416d3dca-66d7-11e6-8463-7e7c6a9f1e68.png">

0xfe commented:
Let me know if this is good, and I'll merge.

Silverwolf90 commented:
Yup this looks great!

0xfe commented:
Thanks for the review. Merged. :-)
